cpp/src/arrow/io/compressed.cc

```diff
@@ -297,10 +297,13 @@ class CompressedInputStream::Impl {
     return Status::OK();
   }
 
-  // Decompress some data from the compressed_ buffer.
-  // Call this function only if the decompressed_ buffer is empty.
+  // Decompress some data from the compressed_ buffer and decompressor_.
+  // Call this function only if the decompressed_ buffer is fully consumed.
   Status DecompressData() {
+    // Currently, compressed_buffer_available() could be 0 in DecompressData()
+    // because `decompressor_` might have its own internal buffer.
```

Comment:
```suggestion
    // compressed_buffer_available() could be 0 here because there might
    // still be some decompressed data left to emit even though the compressed
    // data was entirely consumed (especially if the expansion factor is large)
```